tests/validator/validator.cpp

```diff
@@ -846,6 +846,136 @@ TEST(Validator, validMathMLCiAndCnElementsWithCellMLUnits)
     EXPECT_EQ(size_t(0), v->issueCount());
 }
 
+TEST(Validator, invalidMathMLDegreeOutsideRootOrDiff)
+{
+    const std::string e = "Math has a 'degree' element in a parent 'apply' element; 'degree' is supported for 'root' or 'diff' elements only.";
+    const std::string math =
+        "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
+        "  <apply>\n"
+        "    <eq/>\n"
+        "    <ci>C</ci>\n"
+        "    <apply>\n"
+        "      <degree/>\n"
+        "      <ci>A</ci>\n"
+        "      <ci>B</ci>\n"
+        "    </apply>\n"
+        "  </apply>\n"
+        "</math>\n";
+
+    libcellml::ValidatorPtr v = libcellml::Validator::create();
+    libcellml::ModelPtr m = libcellml::Model::create();
+    libcellml::ComponentPtr c = libcellml::Component::create();
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+    libcellml::VariablePtr v2 = libcellml::Variable::create();
+    libcellml::VariablePtr v3 = libcellml::Variable::create();
+
+    m->setName("modelName");
+    c->setName("componentName");
+    v1->setName("A");
+    v2->setName("B");
+    v3->setName("C");
+    v1->setInitialValue("1.0");
+    v2->setInitialValue("-1.0");
+    v1->setUnits("dimensionless");
+    v2->setUnits("dimensionless");
+    v3->setUnits("dimensionless");
+
+    c->addVariable(v1);
+    c->addVariable(v2);
+    c->addVariable(v3);
+    c->setMath(math);
+    m->addComponent(c);
+
+    v->validateModel(m);
+    EXPECT_EQ(size_t(1), v->errorCount());
+    EXPECT_EQ(e, v->error(0)->description());
+}
+
+TEST(Validator, validMathMLDegreeInsideRoot)
+{
+    auto math = "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
+                "  <apply><eq/>\n"
+                "    <ci>A</ci>\n"
+                "    <apply>\n"
+                "      <root/>\n"
+                "      <degree><ci>B</ci></degree>\n"
+                "      <ci>C</ci>\n"
```

Comment:
If you swap lines 901 and 902, your test will still be valid while it shouldn't be... since your new code in `Validator::ValidatorImpl::validateMathMLElements()` won't be executed. Same, if line 901 was duplicated. The test would still pass while it shouldn't. Etc.

Reply:
The validator we checks against the spec only, and all the spec says is that `degree` must be a child of `root` or `diff`.  It doesn't matter whether the MathML makes sense or not; if the XML says that there's a node called "degree" as a child of "root" then it's ok ... it's not sensible, but, according to the spec, it's ok?  Or are we now introducing stuff that says that the MathML also has to be reasonably constructed in order to be valid? ... because that's an entirely different issue!

Reply:
Agreed that it would be an entirely different issue, which is why I I would just stick with the DTD validation. It has its limitations, but so does this PR, hence I wouldn't bother with it.